### PR TITLE
Add correct column to option

### LINKED ISSUE #153: Add 'correct' (default: false) flag to option
- `Question` has one `QuestionType`
- `Question` has many `Options`
- If `QuestionType` is `MultipleChoice` many  `Options` can have `correct_answer` equals to `true`
- If `QuestionType` is `SingleChoice` only one `Option` can have `correct_answer` equals to `true`
- When `Question` has `QuestionType` equals to `MultipleChoice` and many `options` have the flag `correct` equals to `true`, if user tries to change `QuestionType` to `SingleChoice` it should raises an error.
image

### PR DESCRIPTION
**Pull request by garciajordy**
# CorrectColumn!

**OptionModel:**
- Add column correct with default: false.
- Add validation when create or update options depending on question_type.

**QuestionModel:**
- Add validation for updating question depending on correct options.

**FactoryBot:**
- Add "Single Choice" QuestionType.
- Add "Multiple Choice" QuestionType.
- Add single choice Question.
- Add multiple choice Question.

**Policy:**
- Add GenericPolicy to OptionPolicy.

## Rspec!
**OptionModel:**
- Add validation test for when trying to create a second correct option to single choice question.

**QuestionModel:**
- Add validation test when trying to update to single choice when having multiple correct options.

**Capybara:**
- Add test to see correct changing to true when updating option to true.

_Thanks for your support and time to review my PR!_ 😄 

fix #153

### CODE REVIEW

**Review by edimossilva — CHANGES_REQUESTED** [review]
This is a big PR so we will have many things to change, dont worry about it

File: backend/app/models/option.rb [hunk 1]
```diff
@@ -2,4 +2,16 @@ class Option < ApplicationRecord
...
   belongs_to :question
   validates :name, presence: true
+  validate :validates_correct
+
+  scope :correct, -> { where(correct: true) }
+
+  private
+
+  def validates_correct
+    return unless question
```
[the remark this answers is not in the document]
Comment (garciajordy) [reply backend/app/models/option.rb:12]: ### Done!

File: backend/app/models/question.rb [hunk 2]
```diff
@@ -1,7 +1,15 @@
...
   belongs_to :user
   belongs_to :question_type
   belongs_to :survey, optional: true
   has_many :options, dependent: :destroy
+
+  private
+
+  def validates_options
+    # i18n-tasks-use t('activerecord.errors.models.question.attributes.question_type.cant_change_question_type')
+    errors.add(:question_type, :cant_change_question_type) if options.correct.count > 1 && question_type.name == 'Single Choice'
```
[the remark this answers is not in the document]
Comment (garciajordy) [reply backend/app/models/question.rb:13]: ### Done!

File: backend/spec/factories/questions.rb [hunk 3]
```diff
@@ -5,4 +5,12 @@
     association :question_type
     association :user
   end
+
+  factory :single_choice_question, parent: :question do
+    question_type { QuestionType.find_by(name: 'Single Choice') || create(:question_type_single) }
```
[the remark this answers is not in the document]
Comment (garciajordy) [reply backend/spec/factories/questions.rb:10]: ### Done!

File: backend/spec/models/option_spec.rb [hunk 4]
```diff
@@ -11,5 +12,16 @@
 
   describe 'relationships' do
     it { is_expected.to belong_to(:user).required }
+    it { is_expected.to belong_to(:question).required }
+  end
+
+  describe 'option validation' do
```
[the remark this answers is not in the document]
Comment (garciajordy) [reply backend/spec/models/option_spec.rb:18]: # Done!

File: backend/spec/models/question_spec.rb [hunk 5]
```diff
@@ -15,4 +17,14 @@
   describe 'presence' do
     it { is_expected.to validate_presence_of(:name) }
   end
+
+  describe 'not valid' do
```
[the remark this answers is not in the document]
Comment (garciajordy) [reply backend/spec/models/question_spec.rb:21]: ### Done!

File: backend/app/models/option.rb [hunk 6]
```diff
@@ -2,4 +2,16 @@ class Option < ApplicationRecord
   belongs_to :user
   belongs_to :question
   validates :name, presence: true
+  validate :validates_correct
+
+  scope :correct, -> { where(correct: true) }
```
[the remark this answers is not in the document]
Comment (garciajordy) [reply backend/app/models/option.rb:7]: Done !

**Comment by garciajordy** [comment]
# Changes
**OptionModel:**
- Refactor "validates_correct?" method.
- Add method from the QuestionModel by using a delegate. (was nice to learn more about that!).

**QuestionModel:**
- Create "single_choice?" method.
- Refactor"validates_options" by adding my new method.

**QuestionTypeModel:**
- Add CONST with "Single Choice".
- Add CONST with "Multiple Choice"

## Rspec!
**Option_spec.rb:**
- Add Scope testing for "correct" scope.
- Rename describe for the previous test.

**Question_spec.rb:**
- Rename describe for the previous test.
- Add method test for "single_choice?" method.

_Thanks for taking the time to review my changes!_ 😄

File: backend/app/models/question.rb [hunk 7]
```diff
@@ -1,7 +1,20 @@
...
   has_many :options, dependent: :destroy
+
+  def single_choice?
+    question_type.name == QuestionType::SINGLE_CHOICE
+  end
+
+  private
+
+  def validates_options
+    correct_options = options.correct.count > 1
```
Comment (edimossilva) [inline backend/app/models/question.rb:16]:
```suggestion
    many_correct_options = options.correct.count > 1
```
```suggestion
    correct_options = options.correct.count > 1
```

File: backend/app/models/question.rb [hunk 8]
```diff
@@ -1,7 +1,20 @@
...
+    # i18n-tasks-use t('activerecord.errors.models.question.attributes.question_type.cant_change_question_type')
+    errors.add(:question_type, :cant_change_question_type) if correct_options && single_choice?
```
Comment (edimossilva) [inline backend/app/models/question.rb:18]:
```suggestion
    errors.add(:question_type, :cant_change_question_type) if many_correct_options && single_choice?
```

File: backend/app/models/question.rb [hunk 9]
```diff
@@ -1,7 +1,20 @@
 class Question < ApplicationRecord
   validates :name, presence: true
+  validate :validates_options, on: :update
   belongs_to :user
   belongs_to :question_type
   belongs_to :survey, optional: true
   has_many :options, dependent: :destroy
+
+  def single_choice?
+    question_type.name == QuestionType::SINGLE_CHOICE
```
Comment (edimossilva) [inline backend/app/models/question.rb:10]: this method should be in the `QuestionType` class.
at this class you should use delegate to reference it

File: backend/app/models/question.rb [hunk 10]
```diff
@@ -1,7 +1,20 @@
 class Question < ApplicationRecord
   validates :name, presence: true
+  validate :validates_options, on: :update
```
[the remark this answers is not in the document]
Comment (garciajordy) [reply backend/app/models/question.rb:3]: Ok I think this was an issue before, with my first approach. Now we don't need that anymore! Thanks for pointing that out.

File: backend/app/models/question.rb (hunk 9 above)
Comment (garciajordy) [reply backend/app/models/question.rb:10]: Done!

File: backend/app/models/question.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (garciajordy) [reply backend/app/models/question.rb:3]: Done!

File: backend/spec/models/question_spec.rb [hunk 11]
```diff
@@ -2,6 +2,9 @@
 
 RSpec.describe Question, type: :model do
   let(:question) { create(:question) }
+  let!(:question_multiple) { create(:multiple_choice_question) }
+  let!(:question_single) { create(:single_choice_question) }
+  let!(:question_type_single) { QuestionType.find_by(name: QuestionType::SINGLE_CHOICE) || create(:question_type_single) }
```
[the remark this answers is not in the document]
Comment (garciajordy) [reply backend/spec/models/question_spec.rb:7]: Solved!

File: backend/spec/models/question_spec.rb [hunk 12]
```diff
@@ -15,4 +18,21 @@
...
+    before do
+      create_list(:option, 3, correct: true, question_id: question_multiple.id)
+      question_multiple.update(question_type_id: question_type_single.id)
+    end
+
+    it { expect(question_multiple).not_to be_valid }
+    it { expect(question_multiple.errors.full_messages).to match(["Question type Can't change to single choice when having more than one correct option."]) }
+  end
+
+  describe 'methods' do
```
[the remark this answers is not in the document]
Comment (garciajordy) [reply backend/spec/models/question_spec.rb:32]: Done!

**Comment by garciajordy** [comment]
# Changes

**QuestionModel:**
- Remove "single_choice" method.
- Add delegate "single_choice" to QuestionType.
- Change correct_options to many_correct_options.
- Remove "on update" as we don't need this anymore with my new approach.

**QuestionTypeModel:**
- Add the "single_choice" method as it actually is a question_type method.

**OptionModel:**
- Remove delegate as I can access it from the question association.

### Factory
**QuestionFactory:**
- Use find_or_create_by instead of using an "OR" statement.

### Rspec
**Question_spec:**
- Use find_or_create_by when assigning question_type.
- Remove describe "Methods".

_Thanks for all the hints and requested changes!_ 
_Better for the App and better for my progress!_ 😄

File: backend/spec/factories/questions.rb [hunk 13]
```diff
@@ -5,4 +5,12 @@
     association :question_type
     association :user
   end
+
+  factory :single_choice_question, parent: :question do
+    question_type { QuestionType.find_by(name: QuestionType::SINGLE_CHOICE) || create(:question_type_single) }
```
[the remark this answers is not in the document]
Comment (garciajordy) [reply backend/spec/factories/questions.rb:10]: done!

File: backend/app/models/question.rb (hunk 8 above)
Comment (garciajordy) [reply backend/app/models/question.rb:18]: done!

File: backend/app/models/question.rb (hunk 7 above)
Comment (garciajordy) [reply backend/app/models/question.rb:16]: done!

**Review by edimossilva — APPROVED** [review]
LGTM, Nice!

**Pull request merged**